source/core/Events.cpp

```diff
@@ -1064,14 +1064,45 @@ VIREO_FUNCTION_SIGNATUREV(WaitForEventsAndDispatch, WaitForEventsParamBlock)
-                        // When we support multiple events with a shared event case, this needs to be smarter
-                        // and copy only common data, using a custom emit proc (which could verify types) and
-                        // generate custom snippets for each event data node cluster.
-                        eventData.eventDataType->CopyData(eventData.pEventData, (AQBlock1*)esEventDataNode + commonDataSize);
+                    const UInt32 numCommonElements = EventCommonData::GetNumberOfCommonElements();
+                    const UInt32 numDestElements = esEventDataNodeType->SubElementCount();
+                    const UInt32 numSourceElements = eventData.eventDataType->SubElementCount();
+                    UInt32 destElementIndex = numCommonElements, sourceElementIndex = 0;
+                    if (numDestElements == numCommonElements + 1
+                        && esEventDataNodeType->GetSubElement(destElementIndex)->IsA(eventData.eventDataType)) {
+                        // Special case, all non-commmon data is encapsulated in a single cluster, for example, when
+                        // User Event use the same cluster type which defined the user event to be to fire it and
+                        // as a data element after the common portion (without inlining its fields).
+                        TypeRef destElemType = esEventDataNodeType->GetSubElement(destElementIndex);
+                        eventData.eventDataType->CopyData((AQBlock1*)eventData.pEventData,
+                                                          (AQBlock1*)esEventDataNode + destElemType->ElementOffset());
+                    } else {
+                        // Loop through all elements in the destination event data cluster, and look for the matching
+                        // source element in the fired event data.  We need to copy one element at a time because the
+                        // alignment of the source and dest might not be the same because of the precense of leading
+                        // common fields, and also to handle shared event cases which only expose event data common
+                        // to all sources.
+                        // Note, although this looks like an O(N^2) nested loop, it's actually O(N+M), the sum of
+                        // the number of fields in the source (fired event data) and dest (cluster).
+                        while (destElementIndex < numDestElements) {
+                            // (TODO(spathiwa) We could use a custom emit proc to verify types and generate
+                            // custom snippets for each event data node cluster so it doesn't have to check at run-time.)
+                            TypeRef destElemType = esEventDataNodeType->GetSubElement(destElementIndex);
+                            IntIndex destOffset = destElemType->ElementOffset();
+                            while (sourceElementIndex < numSourceElements) {
```

Comment:
Do we always assume that the order of data in the source and destination must be the same? That way we don't have to reset sourceElementIndex before starting this loop?

Reply:
Yes, this is required. NCG does the same thing.